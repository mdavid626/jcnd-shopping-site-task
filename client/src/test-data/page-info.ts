import { PageInfo } from '../types/page-info';

export const testPageInfo1: PageInfo = {
  page: 1,
  numberOfPages: 5,
};
