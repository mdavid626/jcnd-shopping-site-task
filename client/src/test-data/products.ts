import { Product, ProductsQueryResult } from '../types/products';

export const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  price: 46.36,
  inStock: 5,
};

export const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  price: 25.36,
  inStock: 3,
};

export const testProductQueryResult1: ProductsQueryResult = {
  products: {
    nodes: [testProduct1, testProduct2],
    pageInfo: {
      page: 1,
      numberOfPages: 2,
    },
  },
};
