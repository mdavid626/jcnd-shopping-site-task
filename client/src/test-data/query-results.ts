export const productsQueryResult = {
  data: {
    products: {
      nodes: [
        {
          id: '63ea6485732ff5cf8b364431',
          priceInCents: 9699,
          name: 'Broccoli',
          inStock: 466,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364432',
          priceInCents: 3434,
          name: 'Broccolini',
          inStock: 223,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364433',
          priceInCents: 3692,
          name: 'Brussels Sprouts',
          inStock: 12,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364434',
          priceInCents: 8839,
          name: 'Cabbage',
          inStock: 200,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364435',
          priceInCents: 6828,
          name: 'Carrots',
          inStock: 151,
          __typename: 'Product',
        },
      ],
      pageInfo: { page: 1, numberOfPages: 15, __typename: 'PageInfo' },
      __typename: 'ProductsQueryResult',
    },
  },
};

export const placeOrderQueryResult1 = {
  data: { placeOrder: '63ede5e031f3fc4012a0951b' },
};
