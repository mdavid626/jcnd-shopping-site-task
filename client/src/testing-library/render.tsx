/* istanbul ignore file */
import {
  ApolloClient,
  ApolloProvider,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { InitialEntry, Location } from '@remix-run/router';
import { Queries, queries } from '@testing-library/react';
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
  RenderResult,
  RenderHookResult,
} from '@testing-library/react';
import React, { ReactElement, useMemo, useState } from 'react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';
import { NavigateFunction } from 'react-router/dist/lib/hooks';
import ShoppingCartContext from '../contexts/shopping-cart-context/shopping-cart-context';
import {
  ShoppingCartContextValue,
  ShoppingCartItem,
} from '../types/shopping-cart';

type SimpleRouter = {
  location?: Location;
  navigate?: NavigateFunction;
};

type RouterResult = {
  router: SimpleRouter;
};

type ApolloClientResult = {
  apolloClient: ApolloClient<NormalizedCacheObject>;
};

type ShoppingCartContextResult = {
  shoppingCartContext: { value?: ShoppingCartContextValue };
};

export const renderWithRouter = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  } as RenderResult & RouterResult;
};

export const renderWithRouterAndQueryClient = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult & ApolloClientResult => {
  const router = {} as SimpleRouter;
  const apolloClient = new ApolloClient({
    uri: 'https://api.github.com/graphql',
    cache: new InMemoryCache(),
  });
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <ApolloProvider client={apolloClient}>
            <MemoryRouter initialEntries={initialEntries}>
              <Children />
            </MemoryRouter>
          </ApolloProvider>
        );
      },
      ...options,
    }),
    router,
    apolloClient,
  } as RenderResult & RouterResult & ApolloClientResult;
};

export const renderHookWithRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialEntries?: InitialEntry[]
): RenderHookResult<Result, Props> & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  };
};

export const renderHookWithQueryClient = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>
): RenderHookResult<Result, Props> & ApolloClientResult => {
  const apolloClient = new ApolloClient({
    uri: 'https://api.github.com/graphql',
    cache: new InMemoryCache(),
  });
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => (
        <ApolloProvider client={apolloClient}>{children}</ApolloProvider>
      ),
      ...options,
    }),
    apolloClient,
  };
};

export const renderHookWithShoppingCart = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialShoppingCart?: ShoppingCartItem[]
): RenderHookResult<Result, Props> & ShoppingCartContextResult => {
  let shoppingCartContext: { value?: ShoppingCartContextValue } = {};
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const [shoppingCart, setShoppingCart] = useState(
          initialShoppingCart || []
        );
        // eslint-disable-next-line react-hooks/rules-of-hooks
        shoppingCartContext.value = useMemo(
          () => ({ shoppingCart, setShoppingCart }),
          [shoppingCart, setShoppingCart]
        );
        return (
          <ShoppingCartContext.Provider value={shoppingCartContext.value}>
            {children}
          </ShoppingCartContext.Provider>
        );
      },
      ...options,
    }),
    shoppingCartContext,
  };
};

export const renderHookWithShoppingCartAndQueryClientAndRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialShoppingCart?: ShoppingCartItem[],
  initialEntries?: string[]
): RenderHookResult<Result, Props> &
  ShoppingCartContextResult &
  ApolloClientResult &
  RouterResult => {
  const router = {} as SimpleRouter;
  const apolloClient = new ApolloClient({
    uri: 'https://api.github.com/graphql',
    cache: new InMemoryCache(),
  });
  let shoppingCartContext: { value?: ShoppingCartContextValue } = {};
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const [shoppingCart, setShoppingCart] = useState(
          initialShoppingCart || []
        );
        // eslint-disable-next-line react-hooks/rules-of-hooks
        shoppingCartContext.value = useMemo(
          () => ({ shoppingCart, setShoppingCart }),
          [shoppingCart, setShoppingCart]
        );
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <ApolloProvider client={apolloClient}>
            <ShoppingCartContext.Provider value={shoppingCartContext.value}>
              <MemoryRouter initialEntries={initialEntries}>
                <Children />
              </MemoryRouter>
            </ShoppingCartContext.Provider>
          </ApolloProvider>
        );
      },
      ...options,
    }),
    shoppingCartContext,
    apolloClient,
    router,
  };
};
