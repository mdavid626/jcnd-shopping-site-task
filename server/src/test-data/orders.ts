import { OrderRequestItem } from '../types/order';

export const testOrderRequestItems1: OrderRequestItem[] = [
  { productId: '63ea6485732ff5cf8b36442c', amount: 2 },
  { productId: '63ea6485732ff5cf8b36442d', amount: 4 },
  { productId: '63ea6485732ff5cf8b36442e', amount: 6 },
];
