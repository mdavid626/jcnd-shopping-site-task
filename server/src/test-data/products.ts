import { Product, ProductsQueryResult } from '../types/product';

const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  category: 'VEGETABLES',
  priceInCents: 123,
  inStock: 5,
};

const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  category: 'VEGETABLES',
  priceInCents: 456,
  inStock: 3,
};

export const testProductsQueryResult1: ProductsQueryResult = {
  nodes: [testProduct1, testProduct2],
  pageInfo: {
    page: 1,
    numberOfPages: 5,
  },
};
