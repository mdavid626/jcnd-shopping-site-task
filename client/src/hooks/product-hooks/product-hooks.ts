import { ApolloError, gql, useQuery } from '@apollo/client';
import { ProductCategory, ProductsQueryResult } from '../../types/products';

export const GetProductsQuery = gql`
  query GetProductsQuery($category: ProductCategory!, $page: Int!) {
    products(category: $category, page: $page) {
      nodes {
        id
        price
        name
        inStock
      }
      pageInfo {
        page
        numberOfPages
      }
    }
  }
`;

export const useProducts = (
  category: ProductCategory,
  currentPage: number
): [ProductsQueryResult | undefined, ApolloError | undefined] => {
  const { data, error } = useQuery<ProductsQueryResult>(GetProductsQuery, {
    variables: {
      category,
      page: currentPage,
    },
    notifyOnNetworkStatusChange: true,
  });
  return [data, error];
};
