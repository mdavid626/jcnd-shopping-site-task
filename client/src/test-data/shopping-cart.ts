import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
