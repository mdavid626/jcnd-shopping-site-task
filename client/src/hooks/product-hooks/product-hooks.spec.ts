import { waitFor } from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';
import { productsQueryResult } from '../../test-data/query-results';
import { renderHookWithQueryClient } from '../../testing-library/render';
import { useProducts } from './product-hooks';

describe('product-hooks', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });
  afterEach(jest.resetAllMocks);

  describe('useProducts', () => {
    it('should return products', async () => {
      fetchMock.mockResponse(JSON.stringify(productsQueryResult));
      const { result } = renderHookWithQueryClient(() =>
        useProducts('VEGETABLES', 1)
      );
      expect(result.current).toStrictEqual([undefined, undefined]);
      await waitFor(() => {
        expect(result.current).toStrictEqual([expect.any(Object), undefined]);
      });
      expect(result.current[0]).toMatchSnapshot('query result');
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock.mock.calls[0]).toMatchSnapshot('query');
    });

    it('should error when error', async () => {
      fetchMock.mockRejectedValue(new Error('test error'));
      const { result } = renderHookWithQueryClient(() =>
        useProducts('VEGETABLES', 1)
      );
      expect(result.current).toStrictEqual([undefined, undefined]);
      await waitFor(() => {
        expect(result.current).toStrictEqual([undefined, expect.any(Error)]);
      });
    });
  });
});
