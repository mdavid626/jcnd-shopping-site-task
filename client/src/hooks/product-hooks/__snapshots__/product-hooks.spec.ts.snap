// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product-hooks useProducts should return products: query 1`] = `
Array [
  "https://api.github.com/graphql",
  Object {
    "body": "{\\"operationName\\":\\"GetProductsQuery\\",\\"variables\\":{\\"category\\":\\"VEGETABLES\\",\\"page\\":1},\\"query\\":\\"query GetProductsQuery($category: ProductCategory!, $page: NonNegativeInt!) {\\\\n  products(category: $category, page: $page) {\\\\n    nodes {\\\\n      id\\\\n      priceInCents\\\\n      name\\\\n      inStock\\\\n      __typename\\\\n    }\\\\n    pageInfo {\\\\n      page\\\\n      numberOfPages\\\\n      __typename\\\\n    }\\\\n    __typename\\\\n  }\\\\n}\\"}",
    "headers": Object {
      "accept": "*/*",
      "content-type": "application/json",
    },
    "method": "POST",
    "signal": AbortSignal {},
  },
]
`;

exports[`product-hooks useProducts should return products: query result 1`] = `
Object {
  "products": Object {
    "__typename": "ProductsQueryResult",
    "nodes": Array [
      Object {
        "__typename": "Product",
        "id": "63ea6485732ff5cf8b364431",
        "inStock": 466,
        "name": "Broccoli",
        "priceInCents": 9699,
      },
      Object {
        "__typename": "Product",
        "id": "63ea6485732ff5cf8b364432",
        "inStock": 223,
        "name": "Broccolini",
        "priceInCents": 3434,
      },
      Object {
        "__typename": "Product",
        "id": "63ea6485732ff5cf8b364433",
        "inStock": 12,
        "name": "Brussels Sprouts",
        "priceInCents": 3692,
      },
      Object {
        "__typename": "Product",
        "id": "63ea6485732ff5cf8b364434",
        "inStock": 200,
        "name": "Cabbage",
        "priceInCents": 8839,
      },
      Object {
        "__typename": "Product",
        "id": "63ea6485732ff5cf8b364435",
        "inStock": 151,
        "name": "Carrots",
        "priceInCents": 6828,
      },
    ],
    "pageInfo": Object {
      "__typename": "PageInfo",
      "numberOfPages": 15,
      "page": 1,
    },
  },
}
`;
